### PR TITLE
[BOOKINGSG-9293][ZZ] migrate table to linaria

### PR DESCRIPTION
**Pull request by ziggyzet**
**Checklist**

-   [X] Migrated the component styles
    -   [X] `className` is chained correctly with `clsx`
    -   [X] User style prop is set as CSS variable
-   [X] Changes follow the project guidelines in [CONVENTIONS_V4.md](https://github.com/LifeSG/react-design-system/blob/master/CONVENTIONS_V4.md)
~-   [] Updated Storybook documentation~
-   [X] Added/updated unit tests
-   [X] Added visual tests
~-   [ ] Listed breaking changes~

### CODE REVIEW

File: src/table/table.tsx [hunk 1]
```diff
@@ -17,45 +12,88 @@ Head.displayName = "Table.Head";
...
-        </TableWrapper>
+    ({
+        children,
+        className,
+        ...props
+    }: React.TableHTMLAttributes<HTMLTableElement>) => (
+        <div className={styles.tableWrapper}>
+            <table
+                className={clsx(styles.tableComponent, className)}
+                {...props}
```
Comment (qroll) [inline src/table/table.tsx]: these could reuse `Container` and `TableEl`, so you don't have to repeat the element and class declarations

File: tests/table/table.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        );
+
+        expect(
+            screen.getByRole("columnheader", { name: "Name" })
+        ).toBeInTheDocument();
+        expect(
+            screen.getByRole("columnheader", { name: "Status" })
+        ).toBeInTheDocument();
+        expect(screen.getByText("John Doe")).toBeInTheDocument();
+        expect(screen.getByText("Active")).toBeInTheDocument();
```
Comment (qroll) [inline tests/table/table.spec.tsx]: can use `getByRole("cell")` here as well

File: e2e/tests/components/table/table.e2e.spec.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                    - row "Carol Wong carol@example.com Pending":
+                      - cell "Carol Wong"
+                      - cell "carol@example.com"
+                      - cell "Pending"
+            `);
+        });
+
+        test("Row hover", async ({ story }) => {
+            await story.locators.row1.hover();
+            await compareScreenshot(story, "hover-row", {
```
Comment (qroll) [inline e2e/tests/components/table/table.e2e.spec.ts]: since the name of the test already describes it as row hover, the file name can be simplified to `state`

File: e2e/tests/components/table/table.e2e.spec.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            await compareScreenshot(story, "mount");
+        });
+
+        test("Table can be scrolled horizontally", async ({ story }) => {
+            await story.locators.tableWrapper.hover();
+            await story.page.mouse.wheel(200, 0);
+            const scrollLeft = await story.locators.tableWrapper.evaluate(
+                (el) => el.scrollLeft
+            );
+            expect(scrollLeft).toBeGreaterThan(0);
```
Comment (qroll) [inline e2e/tests/components/table/table.e2e.spec.ts]: not needed since we have the screenshot

File: e2e/tests/components/table/table.e2e.spec.ts
Comment (qroll) [inline e2e/tests/components/table/table.e2e.spec.ts:1]: also include a visual test for vertical overflow

File: e2e/tests/components/table/table.e2e.spec.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        test("Default table (dark mode)", async ({ story }) => {
+            await compareScreenshot(story, "mount");
+        });
+
+        test("Row hover (dark mode)", async ({ story }) => {
+            await story.locators.row1.hover();
+            await compareScreenshot(story, "hover-row", {
+                locator: story.locators.table,
+            });
+        });
```
Comment (qroll) [inline e2e/tests/components/table/table.e2e.spec.ts:87]: there's no hover styling built into the table so these are not needed

**Review by qroll — CHANGES_REQUESTED** [review]

File: src/table/table.styles.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    &:where(&:last-child) {
+        padding-right: ${Spacing["spacing-24"]};
+    }
+`;
+
+export const bodyRow = css`
+    background-color: ${Colour["bg"]};
+    border-top: ${Border["width-010"]} ${Border["solid"]} ${borderColor};
+`;
```
Comment (zhaoyanxzy) [inline src/table/table.styles.ts:65]: i'm expecting `bodyRow` to have a :hover class as per the e2e "hover-row" but doesn't exist. The screenshots for hover looks the same as those of defaults too.

File: src/table/table.styles.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+// =============================================================================
+// STYLES
+// =============================================================================
+export const tableWrapper = css`
+    overflow: auto;
+    border: ${Border["width-010"]} ${Border["solid"]} ${borderColor};
+    border-radius: ${Radius["md"]};
+
+    // Hide scrollbar
```
Comment (zhaoyanxzy) [inline src/table/table.styles.ts]: is this comment needed? the comment syntax is not correct for css

File: src/table/table.tsx [hunk 8]
```diff
@@ -17,45 +12,88 @@ Head.displayName = "Table.Head";
...
-    ({ children, ...props }: React.TableHTMLAttributes<HTMLTableElement>) => (
-        <TableWrapper>
-            <TableComponent {...props}>{children}</TableComponent>
-        </TableWrapper>
+    ({
+        children,
+        className,
+        ...props
+    }: React.TableHTMLAttributes<HTMLTableElement>) => (
+        <div className={styles.tableWrapper}>
```
Comment (zhaoyanxzy) [inline src/table/table.tsx]: here we are not allowing consumers to pass in custom stylings

File: src/table/table.tsx [hunk 9]
```diff
@@ -17,45 +12,88 @@ Head.displayName = "Table.Head";
...
+        {children}
+    </td>
 );
 Cell.displayName = "Table.Cell";
 
 const Header = ({
     children,
+    className,
     ...props
 }: React.ThHTMLAttributes<HTMLTableCellElement>) => (
```
Comment (ghazwan-gt) [inline src/table/table.tsx:50]:
```suggestion
}: React.ThHTMLAttributes<HTMLHeaderCellElement>) => (
```

File: e2e/tests/components/table/table.e2e.spec.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        row2: Locator;
+        row3: Locator;
+    };
+
+    constructor(page: Page) {
+        super(page);
+
+        this.locators = {
+            table: page.getByTestId("table"),
+            tableWrapper: page.getByTestId("table").locator(".."),
```
Comment (ghazwan-gt) [inline e2e/tests/components/table/table.e2e.spec.ts]: Instead having this kind of locator, lets have a wrapper provided with a testid instead

File: e2e/tests/components/table/table.e2e.spec.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            row1: page.getByTestId("row-1"),
+            row2: page.getByTestId("row-2"),
+            row3: page.getByTestId("row-3"),
```
Comment (ghazwan-gt) [inline e2e/tests/components/table/table.e2e.spec.ts]: These two are not used, lets remove them.

File: src/table/table.tsx (hunk 8 above)
Comment (ziggyzet) [reply src/table/table.tsx]: wrapper was never meant to receive custom stylings in the first place. so no need to add in a prop for it

File: src/table/table.tsx (hunk 9 above)
Comment (ghazwan-gt) [reply src/table/table.tsx:50]: Turned out `HTMLTableHeaderCellElement` is already deprecated, thanks for pointing this out @ziggyzet

File: src/table/table.styles.ts (hunk 6 above)
Comment (ziggyzet) [reply src/table/table.styles.ts:65]: will delete the test for this

File: src/table/table.styles.ts (hunk 7 above)
Comment (ziggyzet) [reply src/table/table.styles.ts]: will update the syntax!

File: src/table/table.tsx (hunk 9 above)
Comment (ziggyzet) [reply src/table/table.tsx:50]: type is deprecated, will keep the existing type

**Review by ghazwan-gt — APPROVED** [review]

File: src/table/table.tsx (hunk 8 above)
Comment (qroll) [reply src/table/table.tsx]: if styling on the container is also needed, consumer can compose the two standalone components \o/

File: src/table/table.tsx [hunk 12]
```diff
@@ -17,45 +12,92 @@ Head.displayName = "Table.Head";
...
-            <TableComponent {...props}>{children}</TableComponent>
-        </TableWrapper>
+    ({
+        children,
+        className,
+        ...props
+    }: React.TableHTMLAttributes<HTMLTableElement>) => (
+        <Container>
+            <TableEl
+                className={clsx(styles.tableComponent, className)}
```
Comment (qroll) [inline src/table/table.tsx]: simplify since the classname is already applied on the inner component. no need to destructure className again

```suggestion
            <TableEl
```

File: e2e/nextjs-app/src/app/components/table/table.module.css [hunk 13]
```diff
@@ -0,0 +1,8 @@
+.table-overflow {
+    width: 400px;
+}
+
+.table-vertical-overflow {
+    max-height: 200px;
+    overflow-y: auto;
```
Comment (qroll) [inline e2e/nextjs-app/src/app/components/table/table.module.css]: is this overflow property needed again? container has the overflow property set already

**Review by qroll — APPROVED** [review]
